lib/luaxtable.c

```diff
@@ -0,0 +1,234 @@
+/*
+* Copyright (c) 2024 Mohammad Shehar Yaar Tausif <sheharyaar48@gmail.com>
+*
+* Permission is hereby granted, free of charge, to any person obtaining
+* a copy of this software and associated documentation files (the
+* "Software"), to deal in the Software without restriction, including
+* without limitation the rights to use, copy, modify, merge, publish,
+* distribute, sublicense, and/or sell copies of the Software, and to
+* permit persons to whom the Software is furnished to do so, subject to
+* the following conditions:
+*
+* The above copyright notice and this permission notice shall be
+* included in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/module.h>
+#include <linux/version.h>
+#include <linux/types.h>
+#include <linux/netfilter.h>
+#include <linux/netfilter/x_tables.h>
+
+#include <lua.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+typedef enum luaxtable_type_e {
+	LUAXTABLE_TMATCH,
+	LUAXTABLE_TTARGET,
+} luaxtable_type_t;
+
+typedef struct luaxtable_s {
+	lunatik_object_t *runtime;
+	union {
+		struct xt_match *match;
+		struct xt_target *target;
+	};
+	luaxtable_type_t type;
+} luaxtable_t;
+
+static void luaxtable_release(void *private);
+
+#define luaxtable_setinteger(L,hook, field) 			\
+do {								\
+	lunatik_checkfield(L, 1, #field, LUA_TNUMBER);		\
+	hook->field = lua_tointeger(L, -1);			\
+} while (0)
+
+static inline const char * luaxtable_checkname(lua_State *L, int idx)
+{
+        const char *name;
+        size_t len;
+
+        lunatik_checkfield(L, idx, "name", LUA_TSTRING);
+        name = lua_tolstring(L, -1, &len);
+        if (len >= XT_EXTENSION_MAXNAMELEN)
+                luaL_error(L, "name is too long");
+        return name;
+}
+
+static inline luaxtable_t *luaxtable_new(lua_State *L, const lunatik_class_t *cls, int hook)
```

Comment:
you don't need to pass class around.. you can assume one class per file.. also, let's always use parametrize the stack index..